string center_line
string goal_line
string boundary_line
string undefined_lines